File: benchmarks/python/parse_user.py
from dataclasses import dataclass
from typing import Any


@dataclass
class User:
    name: str
    age: int


@dataclass
class ParseOk:
    value: User


@dataclass
class ParseErr:
    error: str


def parse_user(input: Any) -> ParseOk | ParseErr:
    if not isinstance(input, dict):
        return ParseErr("expected object")
    name = input.get("name")
    if not isinstance(name, str):
        return ParseErr("name: expected string")
    age = input.get("age")
    if not isinstance(age, (int, float)):
        return ParseErr("age: expected number")
    return ParseOk(User(name=name, age=age))

File: benchmarks/rust/parse_user.rs
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Deserialize)]
pub struct User {
    pub name: String,
    pub age: u32,
}

pub fn parse_user(input: &Value) -> Result<User, String> {
    let obj = input.as_object().ok_or_else(|| "expected object".to_string())?;
    let name = obj
        .get("name")
        .and_then(|v| v.as_str())
        .ok_or_else(|| "name: expected string".to_string())?
        .to_string();
    let age = obj
        .get("age")
        .and_then(|v| v.as_u64())
        .ok_or_else(|| "age: expected number".to_string())? as u32;
    Ok(User { name, age })
}
